### PR TITLE
Fix check failure when DisconnectClient happens before AnnounceWorkerPort

### LINKED ISSUE #24070: [Core] Crash in worker_pool  RemoveWorker
### What happened + What you expected to happen

When we debug shuffle_1tb_5000_partitions we encounter following error looks pretty bad

```
[2m[33m(raylet)[0m [2022-04-18 19:37:33,425 E 234 234] (raylet) worker_pool.cc:502: Some workers of the worker process(2654) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet)[0m [2022-04-18 19:37:33,460 E 234 234] (raylet) worker_pool.cc:502: Some workers of the worker process(2655) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet)[0m [2022-04-18 19:37:33,493 E 234 234] (raylet) worker_pool.cc:502: Some workers of the worker process(2656) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[36m(raylet, ip=172.31.82.154)[0m Spilled 9212 MiB, 239997 objects, write throughput 326 MiB/s.
[2m[33m(raylet)[0m [2022-04-18 19:37:33,527 E 234 234] (raylet) worker_pool.cc:502: Some workers of the worker process(2659) have not registered within the timeout. The process is still alive, probably it's hanging during start.
[2m[33m(raylet)[0m [2022-04-18 19:37:33,659 C 234 234] (raylet) worker_pool.cc:1373:  Check failed: RemoveWorker(io_worker_state.registered_io_workers, worker) 
[2m[33m(raylet)[0m *** StackTrace Information ***
[2m[33m(raylet)[0m     ray::SpdLogMessage::Flush()
[2m[33m(raylet)[0m     ray::RayLog::~RayLog()
[2m[33m(raylet)[0m     ray::raylet::WorkerPool::DisconnectWorker()
[2m[33m(raylet)[0m     ray::raylet::NodeManager::DisconnectClient()
[2m[33m(raylet)[0m     ray::raylet::NodeManager::ProcessDisconnectClientMessage()
[2m[33m(raylet)[0m     ray::raylet::NodeManager::ProcessClientMessage()
[2m[33m(raylet)[0m     std::_Function_handler<>::_M_invoke()
[2m[33m(raylet)[0m     ray::ClientConnection::ProcessMessage()
[2m[33m(raylet)[0m     EventTracker::RecordExecution()
[2m[33m(raylet)[0m     boost::asio::detail::read_op<>::operator()()
[2m[33m(raylet)[0m     boost::asio::detail::reactive_socket_recv_op<>::do_complete()
[2m[33m(raylet)[0m     boost::asio::detail::scheduler::do_run_one()
[2m[33m(raylet)[0m     boost::asio::detail::scheduler::run()
[2m[33m(raylet)[0m     boost::asio::io_context::run()
[2m[33m(raylet)[0m     main
[2m[33m(raylet)[0m     __libc_start_main
[2m[33m(raylet)[0m 
[2m[36m(raylet, ip=172.31.80.133)[0m Spilled 9135 MiB, 237997 objects, write throughput 285 MiB/s.
```

https://console.anyscale.com/o/anyscale-internal/projects/prj_SVFGM5yBqK6DHCfLtRMryXHM/clusters/ses_Tir3ZCQ1JMWWMLS7ALB7WeAZ?command-history-section=command_history

### Versions / Dependencies

latest master

### Reproduction script

run shuffle 5000 partition nightly test

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
We have several issues if DisconnectClient happens before AnnounceWorkerPort:
- Check failure for removing io worker from `registered_io_workers` since the io worker is only added to that set after AnnounceWorkerPort.
- num_starting_(io)_workers is not decremented.

## Related issue number
Closes #24070

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jjyao** [comment]
Not 100% sure this is the exact cause of the nightly test failure since it's a bit hard to reproduce but this is what I found by reading the code and the stack trace.

File: src/ray/raylet/worker_pool.cc [hunk 1]
```diff
@@ -1352,12 +1352,15 @@ void WorkerPool::PrestartWorkers(const TaskSpecification &task_spec,
...
+void WorkerPool::DisconnectWorker(const std::shared_ptr<WorkerInterface> &worker,
                                   rpc::WorkerExitType disconnect_type) {
   MarkPortAsFree(worker->AssignedPort());
   auto &state = GetStateForLanguage(worker->GetLanguage());
   auto it = state.worker_processes.find(worker->GetStartupToken());
   if (it != state.worker_processes.end()) {
+    if (!RemoveWorker(it->second.alive_started_workers, worker)) {
+      it->second.num_starting_workers--;
+    }
     it->second.alive_started_workers.erase(worker);
```
Comment (stephanie-wang) [inline src/ray/raylet/worker_pool.cc:1364]: Remove this line?

**Review by stephanie-wang — APPROVED** [review]
Does the test pass with this change?

File: src/ray/raylet/worker_pool.cc (hunk 1 above)
Comment (jjyao) [reply src/ray/raylet/worker_pool.cc]: oops.

**Comment by jjyao** [comment]
> Does the test pass with this change?

Yes.

**Review by scv119 — APPROVED** [review]

File: src/ray/raylet/worker_pool.cc [hunk 2]
```diff
@@ -1352,13 +1352,15 @@ void WorkerPool::PrestartWorkers(const TaskSpecification &task_spec,
...
 
-bool WorkerPool::DisconnectWorker(const std::shared_ptr<WorkerInterface> &worker,
+void WorkerPool::DisconnectWorker(const std::shared_ptr<WorkerInterface> &worker,
                                   rpc::WorkerExitType disconnect_type) {
   MarkPortAsFree(worker->AssignedPort());
   auto &state = GetStateForLanguage(worker->GetLanguage());
   auto it = state.worker_processes.find(worker->GetStartupToken());
   if (it != state.worker_processes.end()) {
-    it->second.alive_started_workers.erase(worker);
+    if (!RemoveWorker(it->second.alive_started_workers, worker)) {
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:1361]: Can you add a comment on this block? I am a little bit confused here (isn't it supposed to be `if (RemoveWorker)`)?

File: src/ray/raylet/worker_pool.cc [hunk 3]
```diff
@@ -1370,8 +1372,11 @@ bool WorkerPool::DisconnectWorker(const std::shared_ptr<WorkerInterface> &worker
   if (IsIOWorkerType(worker->GetWorkerType())) {
     auto &io_worker_state =
         GetIOWorkerStateFromWorkerType(worker->GetWorkerType(), state);
-    RAY_CHECK(RemoveWorker(io_worker_state.registered_io_workers, worker));
-    return RemoveWorker(io_worker_state.idle_io_workers, worker);
+    if (!RemoveWorker(io_worker_state.started_io_workers, worker)) {
```
Comment (rkooo567) [inline src/ray/raylet/worker_pool.cc:1375]: Same comment as above

File: src/ray/raylet/worker_pool.cc (hunk 2 above)
Comment (rkooo567) [reply src/ray/raylet/worker_pool.cc:1361]: I assume it means "if the worker is not started yet, it means it is starting, so we should subtract it?"

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Btw, not that I think it is a better direction, but why is it splitted into 2 RPCs? Feel like AnnounceWorkerPort and RegisterClient could be combined into one... but I might be missing some edge cases

**Comment by jjyao** [comment]
Yup, that's something I'm thinking as well. My understanding is that RegisterClient rpc is used to get the port that core worker should listen on and the follow-up AnnounceWorkerPort is when the port is actually listened. But I think the port can be given when raylet starts core worker process (e.g. --port).

It's a bigger change so we should merge this first for 1.13.

**Comment by rkooo567** [comment]
Should we create an issue for that? This separate RPCs seems to hard to reason (and ordering is also not guaranteed, although I guess it is rare in a local node env like this), and I feel like combining it will make it much easier to avoid edge cases like this.

**Comment by jjyao** [comment]
> Should we create an issue for that? This separate RPCs seems to hard to reason (and ordering is also not guaranteed, although I guess it is rare in a local node env like this), and I feel like combining it will make it much easier to avoid edge cases like this.

Created one issue: https://github.com/ray-project/ray/issues/24432. The order is guaranteed here since it's through the same socket directly not grpc.

**Pull request merged**